docs/code-generation-design.md

```diff
@@ -0,0 +1,251 @@
+   // the node.
+	Name        string
+	// YANGDetails stores details of the node from the original
+	// YANG schema, such that some characteristics can be accessed
+	// by the code generation process. Only details that are
+	// directly required are provided.
+	YANGDetails YANGNodeDetails
+	// Type describes the type of node that the leaf represents,
+	// allowing for container, list, leaf and leaf-list entries
+	// to be distinguished.
+	// In the future it can be used to store other node types that
+	// form a direct child of a subtree node.
+	Type        NodeType
+	// LangType describes the type that the node should be given in
+	// the output code, using the output of the language-specific
+	// type mapping provided by calling the LangMapper interface.
+	LangType    *MappedType
+	// MapPaths describes the paths that the output node should
+	// be mapped to in the output code - these annotations can be
+	// used to annotation the output code with the field(s) that it
+	// corresponds to in the YANG schema.
+	MapPaths    [][]string
+}
+
+// YANGNodeDetails stores the YANG-specific details of a node
+// within the schema.
+type YANGNodeDetails struct {
+	// Name is the name of the node from the YANG schema.
+	Name    string
+	// Default represents the 'default' value directly
+	// specified in the YANG schema.
+	Default string
+	// Module stores the name of the module that instantiates
+	// the node.
+	Module  string
+	// Path specifies the complete YANG schema node path.
+	Path    []string
+}
+```
+
+#### Definitions for Enumerated Types
+
+YANG has a number of different types of enumerated values - particularly, `identity` and `enumeration` statements. Generated code may choose to treat these differently. For example, for an `enumeration` leaf - some languages may choose to use an embedded enumeration (e.g., protobuf can use a scoped `Enum` within a message). For this reason, some of the provenance of a particular value needs to be exposed to the downstream code generation libraries.
+
+Despite this, a significant amount of pre-parsing can be done by the `ygen` library to pre-process enumerated values prior to being output to code. Particularly:
+
+* Handling of an enumerated value's integer ID to an language-specific name can be created (using the relevant features of the `LangMapper` interface).
+* A reverse mapping between the integer ID and defining module and corresponding YANG identifier can be created.
+
+By performing this pre-processing, the functionality of the code generation library is constrained to determine how each type of enumeration should be output, and creating the language-specific constructs required for mapping (e.g., the enumeration map created in `ygen`'s current Go output).
+
+The IR format for enumerations is defined to be as follows:
+
+```golang
+type EnumeratedValueType int64
+
+const (
+	_ EnumeratedValueType = iota
+	SimpleEnumerationType
+	DerivedEnumerationType
+	UnionEnumerationType
```

Comment:
With the new enum changes, is there now a difference between an enum within a typedef union vs. a not within a typedef union that's worth putting here?

Reminder that the new enum changes effectively moved a portion of enums previously considered to be simple enumerations used within a leaf's union type, to typedef enums that belong to the typedef union under which they're defined. The only difference is that since they're not regular typedef enums, they're distinguished with a `_Enum` suffix.

Reply:
Yes, I think that we probably do need to have different definitions here, an enumeration in a typedef union is going to be a global-scoped type in the generated code, whereas an enumeration that is within a non-typedef union can still be locally scoped.

I've added these in.